File: src/main/java/com/kivicms/test/controllers/CarController.java
package com.kivicms.test.controllers;

import com.kivicms.test.models.Car;
import com.kivicms.test.models.Mark;
import com.kivicms.test.models.Vendor;
import com.kivicms.test.repositories.CarRepository;
import com.kivicms.test.repositories.MarkRepository;
import com.kivicms.test.repositories.VendorRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import java.util.Map;
import java.util.Optional;

@RequestMapping("/car")
@Controller
public class CarController {

    @Autowired
    private MarkRepository markRepository;

    @Autowired
    private VendorRepository vendorRepository;

    @Autowired
    private CarRepository carRepository;

    @GetMapping
    public String index(Map<String, Object> model) {

        Iterable<Car> cars = carRepository.findAll();
        model.put("cars", cars);

        String breadcrumbs[][] = new String[2][2];

        breadcrumbs[0][0] = "/car";
        breadcrumbs[0][1] = "Автомобили";
        breadcrumbs[1][0] = "";
        breadcrumbs[1][1] = "Список";

        model.put("breadcrumbs", breadcrumbs);

        return "car/index";
    }

    @GetMapping("/create")
    public String create(Map<String, Object> model) {

        Iterable<Mark> marks = markRepository.findAll();
        model.put("markList", marks);

        Iterable<Vendor> vendorList = vendorRepository.findAll();
        model.put("vendorList", vendorList);

        Car car = new Car(" ");
        model.put("car", car);

        return "car/create";
    }

    @PostMapping("/create")
    public String add(@RequestParam Vendor vendor, @RequestParam Mark mark, String number) {
        Car car = new Car(vendor, mark, number, 1);
        carRepository.save(car);
        return "redirect:/car";
    }

    @GetMapping("/view/{id}")
    public String view(@PathVariable Integer id, Map<String, Object> model) {
        Car car = carRepository.findById(id).get();
        model.put("car", car);
        return "car/view";
    }

    @GetMapping("/update/{id}")
    public String edit(@PathVariable Integer id, Map<String, Object> model) {
        Car car = carRepository.findById(id).get();

        model.put("car", car);

        Iterable<Mark> marks = markRepository.findAll();
        model.put("markList", marks);

        Iterable<Vendor> vendorList = vendorRepository.findAll();
        model.put("vendorList", vendorList);

        return "car/update";
    }

    @PostMapping("/update/{id}")
    public String update(@PathVariable Integer id, @RequestParam Vendor vendor, @RequestParam Mark mark, String number) {
        Car car = carRepository.findById(id).get();
        car.setId(id);
        car.setVendor(vendor);
        car.setMark(mark);
        car.setNumber(number);
        carRepository.save(car);
        return "redirect:/car/view/" + id;
    }

    @GetMapping("/delete/{id}")
    public String delete(@PathVariable Integer id) {
        carRepository.deleteById(id);
        return "redirect:/car";
    }
}

File: src/main/java/com/kivicms/test/controllers/CarController.kt
package com.kivicms.test.controllers

import com.kivicms.test.models.Breadcrumb
import com.kivicms.test.models.Car
import com.kivicms.test.repositories.CarRepository
import com.kivicms.test.repositories.MarkRepository
import com.kivicms.test.repositories.VendorRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.*

import java.util.ArrayList

@RequestMapping("/car")
@Controller
class CarController {

    @Autowired
    private val markRepository: MarkRepository? = null

    @Autowired
    private val vendorRepository: VendorRepository? = null

    @Autowired
    private val carRepository: CarRepository? = null

    @GetMapping
    fun index(model: MutableMap<String, Any>): String {

        val cars = carRepository!!.findAll()
        model["cars"] = cars

        val breadcrumbs = ArrayList<Breadcrumb>()
        breadcrumbs.add(Breadcrumb("/car", "Автомобили"))
        breadcrumbs.add(Breadcrumb("", "Список"))
        model["breadcrumbs"] = breadcrumbs

        model["pageTitle"] = "Автомобили"
        return "car/index"
    }

    @GetMapping("/create")
    fun create(model: MutableMap<String, Any>): String {

        val marks = markRepository!!.findAll()
        model["markList"] = marks

        val vendorList = vendorRepository!!.findAll()
        model["vendorList"] = vendorList

        val car = Car(" ")
        model["car"] = car

        val breadcrumbs = ArrayList<Breadcrumb>()
        breadcrumbs.add(Breadcrumb("/car", "Автомобили"))
        breadcrumbs.add(Breadcrumb("", "Добавить"))
        model["breadcrumbs"] = breadcrumbs

        model["pageTitle"] = "Добавить автомобиль"
        return "car/create"
    }

    @PostMapping("/create")
    fun add(@RequestParam vendorId: Int?, @RequestParam markId: Int?, number: String): String {
        val vendor = vendorRepository!!.findById(vendorId).get()
        val mark = markRepository!!.findById(markId).get()
        val car = Car(vendor, mark, number, 0)
        carRepository!!.save(car)
        return "redirect:/car"
    }

    @GetMapping("/view/{id}")
    fun view(@PathVariable id: Int?, model: MutableMap<String, Any>): String {
        val car = carRepository!!.findById(id).get()
        model["car"] = car

        val breadcrumbs = ArrayList<Breadcrumb>()
        breadcrumbs.add(Breadcrumb("/car", "Автомобили"))
        breadcrumbs.add(Breadcrumb("", "Просмотр: " + car.number))
        model["breadcrumbs"] = breadcrumbs

        model["pageTitle"] = "Просмотр автомобиля:" + car.number
        return "car/view"
    }

    @GetMapping("/update/{id}")
    fun edit(@PathVariable id: Int?, model: MutableMap<String, Any>): String {
        val car = carRepository!!.findById(id).get()

        model["car"] = car

        val marks = markRepository!!.findAll()
        model["markList"] = marks

        val vendorList = vendorRepository!!.findAll()
        model["vendorList"] = vendorList

        val breadcrumbs = ArrayList<Breadcrumb>()
        breadcrumbs.add(Breadcrumb("/car", "Автомобили"))
        breadcrumbs.add(Breadcrumb("", "Редактирование: " + car.number))
        model["breadcrumbs"] = breadcrumbs

        model["pageTitle"] = "Редактирование автомобиля:" + car.number
        return "car/update"
    }

    @PostMapping("/update/{id}")
    fun update(@PathVariable id: Int?, @RequestParam vendorId: Int?, @RequestParam markId: Int?, number: String): String {
        val vendor = vendorRepository!!.findById(vendorId).get()
        val mark = markRepository!!.findById(markId).get()
        val car = carRepository!!.findById(id).get()
        car.id = id
        car.vendor = vendor
        car.mark = mark
        car.number = number
        carRepository.save(car)
        return "redirect:/car/view/" + id!!
    }

    @GetMapping("/delete/{id}")
    fun delete(@PathVariable id: Int?): String {
        carRepository!!.deleteById(id)
        return "redirect:/car"
    }
}
